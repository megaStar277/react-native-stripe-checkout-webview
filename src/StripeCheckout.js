/* @flow */
import React, { useState } from 'react';
import { Text } from 'react-native';
import { WebView } from 'react-native-webview';

import stripeCheckoutRedirectHTML from './stripeCheckoutRedirectHTML';

type Props = {
  /** Stripe public key */
  stripePublicKey: string,
  /** Stripe Checkout Session input */
  checkoutSessionInput:
    | {
        sessionId: string,
      }
    | {
        clientReferenceId: string,
        successUrl: string,
        cancelUrl: string,
        items?: Array<{ plan: string, quantity: string }>,
      },
  /** Called when the Stripe checkout session completes with status 'success' */
  onSuccess: ({ [key: string]: any, checkoutSessionId?: string }) => any,
  /** Called when the Stripe checkout session completes with status 'cancel' */
  onCancel: ({ [key: string]: any }) => any,
  /** Called when the Stripe checkout session webpage loads successfully */
  onLoadingComplete?: (syntheticEvent: SyntheticEvent) => any,
  /** Extra options */
  options?: {
    /** The loading item is set on the element with id='sc-loading' */
    htmlContentLoading?: string,
    /** The error is set on the element with id='sc-error-message' */
    htmlContentError?: string,
  },
  /** Props passed to the WebView */
  webViewProps?: Object,
  /** Renders the component shown when checkout session is completed */
  renderOnComplete?: () => React$Node,
};

/**
 * StripeCheckoutWebView
 *
 * Handles a full Stripe Checkout journey on react native via webview
 *
 * Important Notes about URLs:
 * - successUrl must have the query string params `?sc_checkout=success&sc_sid={CHECKOUT_SESSION_ID}`
 *   - sc_sid is optional - must be the last param - when passed results in sessionId being passed to the onSuccess function
 * - cancelUrl must have the query string params `?sc_checkout=cancel`
 */
const StripeCheckoutWebView = (props: Props) => {
  const {
    stripePublicKey,
    checkoutSessionInput,
    onSuccess,
    onCancel,
    onLoadingComplete,
    options,
    webViewProps = {},
    renderOnComplete,
  } = props;
  /** Holds the complete URL if exists */
  const [completed, setCompleted] = useState(null);
  /** Holds wether Stripe Checkout has loaded yet */
  const [hasLoaded, setHasLoaded] = useState(false);

  /**
   * Called everytime the URL stats to load in the webview
   *
   * handles completing the checkout session
   */
  const _onLoadStart = (syntheticEvent: SyntheticEvent) => {
    const { nativeEvent } = syntheticEvent;
    const { url: currentUrl } = nativeEvent;
    /** Check and handle checkout state: success */
    if (currentUrl.includes('sc_checkout=success')) {
      const checkoutSessionIdKey = 'sc_sid=';
      const checkoutSessionId = currentUrl
        .substring(currentUrl.indexOf(checkoutSessionIdKey), currentUrl.length)
        /** remove key */
        .replace(checkoutSessionIdKey, '')
        /** remove extra trailing slash */
        .replace('/', '');
      setCompleted(true);
      if (onSuccess) {
        onSuccess({ ...props, checkoutSessionId });
      }
      return;
    }
    /** Check and handle checkout state: cancel */
    if (currentUrl.includes('sc_checkout=cancel')) {
      setCompleted(true);
      if (onCancel) {
        onCancel(props);
      }
    }
    /** call webViewProps.onLoadStart */
    if (webViewProps && webViewProps.onLoadStart) {
      webViewProps.onLoadStart(syntheticEvent);
    }
  };

  /**
   * Called upon URL load complete
   */
  const _onLoadEnd = (syntheticEvent: SyntheticEvent) => {
    const { nativeEvent } = syntheticEvent;
    /** set isLoading to false once the stripe checkout page loads */
    if (!hasLoaded && nativeEvent.url.startsWith('https://checkout.stripe.com') && onLoadingComplete) {
      setHasLoaded(true);
      onLoadingComplete(syntheticEvent);
    }
    /** call webViewProps.onLoadStart */
    if (webViewProps && webViewProps.onLoadEnd) {
      webViewProps.onLoadEnd(syntheticEvent);
    }
  };

  /** If the checkout session is complete -- render the complete content */
  if (completed) {
    return renderOnComplete ? (
      renderOnComplete({ url: completed, ...props })
    ) : (
      <Text>Stripe Checkout session complete.</Text>
    );
  }

  /** Render the WebView holding the Stripe checkout flow */
  return (
    <WebView
      /** pass baseUrl to avoid  `IntegrationError: Live Stripe.js integrations must use HTTPS.` error https://github.com/react-native-community/react-native-webview/issues/1317 */
      baseUrl=""
      originWhitelist={['*']}
      {...webViewProps}
      source={{
        html: stripeCheckoutRedirectHTML(
          stripePublicKey,
          checkoutSessionInput,
          options,
        ),
        ...webViewProps?.source,
      }}
      onLoadStart={_onLoadStart}
      onLoadEnd={_onLoadEnd}
    />
  );
};

export default StripeCheckoutWebView;
