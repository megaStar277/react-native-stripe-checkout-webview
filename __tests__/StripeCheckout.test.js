import React from 'react';
import { shallow } from 'enzyme';
import StripeCheckout from '../src/StripeCheckout';

const render = (overrideProps) => (
  <StripeCheckout
    stripePublicKey={'sp_test_STRIPE_PUBLIC_KEY'}
    checkoutSessionInput={{
      sessionId: 'CHECKOUT_SESSION_ID',
    }}
    {...overrideProps}
  />
);

describe('<StripeCheckout />', () => {
  it('renders props correctly', () => {
    const wrapper = shallow(render());
    expect(wrapper).toMatchSnapshot();
  });

  it('renders props correctly - extra props', () => {
    const wrapper = shallow(render({
      onSuccess: jest.fn(),
      onCancel: jest.fn(),
      onLoadingComplete: jest.fn(),
      renderOnComplete: jest.fn(),
    }));
    expect(wrapper).toMatchSnapshot();
  });

  it('renders props correctly - webViewProps', () => {
    const wrapper = shallow(
      render({
        webViewProps: {
          originWhitelist: ['https://stripe.com'],
        },
      }),
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders props correctly - options', () => {
    const wrapper = shallow(
      render({
        options: {
          htmlContentLoading: '<p>htmlContentLoading test</p>',
          htmlContentError: '<p>htmlContentError test</p>',
        },
      }),
    );
    expect(wrapper).toMatchSnapshot();
  });
});
