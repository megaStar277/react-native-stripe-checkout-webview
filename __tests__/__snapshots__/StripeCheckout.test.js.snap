// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StripeCheckout /> renders props correctly - extra props 1`] = `
<WebView
  baseUrl=""
  cacheEnabled={true}
  javaScriptEnabled={true}
  onLoadEnd={[Function]}
  onLoadStart={[Function]}
  originWhitelist={
    Array [
      "*",
    ]
  }
  source={
    Object {
      "html": "
  <html>
    <head>
      <meta charset=\\"utf-8\\">
      <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
      <title>Stripe Checkout</title>
      <meta name=\\"author\\" content=\\"A-Tokyo\\">  
    </head>
    <body>
      <!-- Display loading content -->
      <h1 id=\\"sc-loading\\">Loading...</h1>
      <!-- Display error content -->
      <div id=\\"sc-error-message\\"></div>
      <!-- Exec JS without blocking dom -->      
      <!-- Load Stripe.js -->
      <script src=\\"https://js.stripe.com/v3\\"></script>
      <!-- Stripe execution script -->
      <script>
        (function initStripeAndRedirectToCheckout () {
          const stripe = Stripe('sp_test_STRIPE_PUBLIC_KEY');
          window.onload = () => {
            // Redirect to Checkout
            stripe.redirectToCheckout({\\"sessionId\\":\\"CHECKOUT_SESSION_ID\\"})
              .then((result) => {
                // Remove loading html
                const loadingElement = document.getElementById('sc-loading');
                if (loadingElement) {
                  loadingElement.outerHTML = '';
                }
                // If redirectToCheckout fails due to a browser or network
                // error, display the localized error message to your customer.
                if (result.error) {
                  const displayError = document.getElementById('sc-error-message');
                  if (displayError) {
                    displayError.textContent = result.error.message;
                  }
                }
              });
          };
        })();
      </script>
    </body>
  </html>
  ",
    }
  }
  useSharedProcessPool={true}
/>
`;

exports[`<StripeCheckout /> renders props correctly - options 1`] = `
<WebView
  baseUrl=""
  cacheEnabled={true}
  javaScriptEnabled={true}
  onLoadEnd={[Function]}
  onLoadStart={[Function]}
  originWhitelist={
    Array [
      "*",
    ]
  }
  source={
    Object {
      "html": "
  <html>
    <head>
      <meta charset=\\"utf-8\\">
      <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
      <title>Stripe Checkout</title>
      <meta name=\\"author\\" content=\\"A-Tokyo\\">  
    </head>
    <body>
      <!-- Display loading content -->
      <p>htmlContentLoading test</p>
      <!-- Display error content -->
      <p>htmlContentError test</p>
      <!-- Exec JS without blocking dom -->      
      <!-- Load Stripe.js -->
      <script src=\\"https://js.stripe.com/v3\\"></script>
      <!-- Stripe execution script -->
      <script>
        (function initStripeAndRedirectToCheckout () {
          const stripe = Stripe('sp_test_STRIPE_PUBLIC_KEY');
          window.onload = () => {
            // Redirect to Checkout
            stripe.redirectToCheckout({\\"sessionId\\":\\"CHECKOUT_SESSION_ID\\"})
              .then((result) => {
                // Remove loading html
                const loadingElement = document.getElementById('sc-loading');
                if (loadingElement) {
                  loadingElement.outerHTML = '';
                }
                // If redirectToCheckout fails due to a browser or network
                // error, display the localized error message to your customer.
                if (result.error) {
                  const displayError = document.getElementById('sc-error-message');
                  if (displayError) {
                    displayError.textContent = result.error.message;
                  }
                }
              });
          };
        })();
      </script>
    </body>
  </html>
  ",
    }
  }
  useSharedProcessPool={true}
/>
`;

exports[`<StripeCheckout /> renders props correctly - webViewProps 1`] = `
<WebView
  baseUrl=""
  cacheEnabled={true}
  javaScriptEnabled={true}
  onLoadEnd={[Function]}
  onLoadStart={[Function]}
  originWhitelist={
    Array [
      "https://stripe.com",
    ]
  }
  source={
    Object {
      "html": "
  <html>
    <head>
      <meta charset=\\"utf-8\\">
      <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
      <title>Stripe Checkout</title>
      <meta name=\\"author\\" content=\\"A-Tokyo\\">  
    </head>
    <body>
      <!-- Display loading content -->
      <h1 id=\\"sc-loading\\">Loading...</h1>
      <!-- Display error content -->
      <div id=\\"sc-error-message\\"></div>
      <!-- Exec JS without blocking dom -->      
      <!-- Load Stripe.js -->
      <script src=\\"https://js.stripe.com/v3\\"></script>
      <!-- Stripe execution script -->
      <script>
        (function initStripeAndRedirectToCheckout () {
          const stripe = Stripe('sp_test_STRIPE_PUBLIC_KEY');
          window.onload = () => {
            // Redirect to Checkout
            stripe.redirectToCheckout({\\"sessionId\\":\\"CHECKOUT_SESSION_ID\\"})
              .then((result) => {
                // Remove loading html
                const loadingElement = document.getElementById('sc-loading');
                if (loadingElement) {
                  loadingElement.outerHTML = '';
                }
                // If redirectToCheckout fails due to a browser or network
                // error, display the localized error message to your customer.
                if (result.error) {
                  const displayError = document.getElementById('sc-error-message');
                  if (displayError) {
                    displayError.textContent = result.error.message;
                  }
                }
              });
          };
        })();
      </script>
    </body>
  </html>
  ",
    }
  }
  useSharedProcessPool={true}
/>
`;

exports[`<StripeCheckout /> renders props correctly 1`] = `
<WebView
  baseUrl=""
  cacheEnabled={true}
  javaScriptEnabled={true}
  onLoadEnd={[Function]}
  onLoadStart={[Function]}
  originWhitelist={
    Array [
      "*",
    ]
  }
  source={
    Object {
      "html": "
  <html>
    <head>
      <meta charset=\\"utf-8\\">
      <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1\\">
      <title>Stripe Checkout</title>
      <meta name=\\"author\\" content=\\"A-Tokyo\\">  
    </head>
    <body>
      <!-- Display loading content -->
      <h1 id=\\"sc-loading\\">Loading...</h1>
      <!-- Display error content -->
      <div id=\\"sc-error-message\\"></div>
      <!-- Exec JS without blocking dom -->      
      <!-- Load Stripe.js -->
      <script src=\\"https://js.stripe.com/v3\\"></script>
      <!-- Stripe execution script -->
      <script>
        (function initStripeAndRedirectToCheckout () {
          const stripe = Stripe('sp_test_STRIPE_PUBLIC_KEY');
          window.onload = () => {
            // Redirect to Checkout
            stripe.redirectToCheckout({\\"sessionId\\":\\"CHECKOUT_SESSION_ID\\"})
              .then((result) => {
                // Remove loading html
                const loadingElement = document.getElementById('sc-loading');
                if (loadingElement) {
                  loadingElement.outerHTML = '';
                }
                // If redirectToCheckout fails due to a browser or network
                // error, display the localized error message to your customer.
                if (result.error) {
                  const displayError = document.getElementById('sc-error-message');
                  if (displayError) {
                    displayError.textContent = result.error.message;
                  }
                }
              });
          };
        })();
      </script>
    </body>
  </html>
  ",
    }
  }
  useSharedProcessPool={true}
/>
`;
